script/modelling/geographic_clustering.jl

```diff
@@ -0,0 +1,107 @@
+    lat=[40.7127,34.0522,41.8781,29.7604,39.9525,33.4483,29.4241,32.7157,32.7766,37.3382,30.2671,39.7684,30.3321,37.7749,39.9611,35.2270,32.7554,42.3314,31.7775,35.1495],
+    lon=[-74.0059,-118.2436,-87.6297,-95.3698,-75.1652,-112.0740,-98.4936,-117.1610,-96.7969,-121.8863,-97.7430,-86.1580,-81.6556,-122.4194,-82.9987,-80.8431,-97.3307,-83.0457,-106.4424,-90.0489])
+
+#' ### Model Specifics
+
+#' We will cluster these 20 cities into 3 different groups and we will assume that the ideal or target population $P$ for a
+#' group is simply the total population of the 20 cities divided by 3:
+
+n = size(cities,1)
+k = 3
+P = sum(cities.population) / k
+
+#' ### Obtaining the distances between each city
+
+#' Let's leverage the *Distances.jl* package to compute the pairwise Haversine distance between each of the cities in our data
+#' set and store the result in a variable we'll call `dm`:
+
+dm = Distances.pairwise(Haversine(6372.8), Matrix(cities[:, [3,4]])', dims=2)
+
+#' Our distance matrix is symmetric so we'll convert it to a `LowerTriangular` matrix so that we can better interpret the
+#' objective value of our model (if we don't do this the total distance will be doubled):
+
+dm = LowerTriangular(dm)
+
+#' ### Build the model
+#' Now that we have the basics taken  care of, we can set up our model, create decision variables, add constraints, and then
+#' solve.
+
+#' First, we'll set up a model that leverages the [Cbc](https://github.com/coin-or/Cbc) solver. Next, we'll set up a binary
+#' variable $x_{i,k}$ that takes the value $1$ if city $i$ is in group $k$ and $0$ otherwise. Each city must be in a group, so
+#' we'll add the constraint $\sum_kx_{i,k} = 1$ for every $i$.
+
+model = Model(Cbc.Optimizer)
+
+@variable(model, x[1:n, 1:k], Bin)
+
+for i in 1:n
+    @constraint(model, sum(x[i,:]) == 1)
+end
+
+#'The total population of a group $k$ is $Q_k = \sum_ix_{i,k}q_i$ where $q_i$ is simply the $i$th value from the `population`
+#' column in our `cities` DataFrame. Let's add constraints so that $\alpha \leq (Q_k - P) \leq \beta$. We'll set $\alpha$
+#' equal to -2,500,000 and $\beta$ equal to 2,500,000. By adjusting these thresholds you'll find that there is a tradeoff
+#' between having relatively even populations between groups and having geographically close cities within each group. In
+#' other words, the larger the absolute values of $\alpha$ and $\beta$, the closer together the cities in a group will be but
+#' the variance between the group populations will be higher.
+
+α = -2_500_000
+β = 2_500_000
+
+for i in 1:k
+    @constraint(model, (x' * cities.population)[i] - P <= β)
+    @constraint(model, (x' * cities.population)[i] - P >= α)
+end
+
+#' We can now add an objective to our model which will simply be to minimize the dot product of $z$ and our distance matrix,
+#' `dm`. We can then call `optimize!` and review the results.
+
```

Comment:
You forgot to define `z` in the notebook.
```
@variable(model, z[1:n,1:n], Bin)

for k in 1:k, i in 1:n, j in 1:n
    @constraint(model, z[i,j] >= x[i,k] + x[j,k] - 1)
end
```

Reply:
Good catch. Just fixed it.